Add command aliases

Trenly opened the pull request:
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Are you working against an Issue?
  * Resolves #1299
  * Resolves #2303 
  * Resolves #1978
  * Resolves #1026
  * Resolves #380


I have tested that this is functional, and that alias collisions do not occur when commands are at different levels. E.g - `winget add <params>` does not conflict with `winget source add <params>`.

I need some guidance on how to implement the unit tests for this - mainly just how to modify the `EnsureCommandConsistency` test. I presume that the command consistency will need to check against command full names, and that command aliases will have to have a method added to get the alias full name, but I'd like some help and guidance before making those changes


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2390)

Trenly commented:
@ryfu-msft / @JohnMcPMS - Do you have any recommendations on how I should modify the `EnsureCommandConsistency` test?

JohnMcPMS commented:
@denelon for this potential change to the user interface.  Just want to make sure that he is aware and agreeable to the aliases being added.

I assume that many of these are less about the amount of typing and more about offering more verbs for a more intuitive experience.

There has been a lack of rigor (mostly from me) about completion, with many new features not being added.  I'm not going to hold you to it, but now that I've started thinking about it I see it more.

I would think that the tests should be able to steal from the way that the aliases for arguments are verified:
https://github.com/microsoft/winget-cli/blob/888b4ed8f4f7d25cb05a47210e083fe29348163b/src/AppInstallerCLITests/Command.cpp#L91-L93

Trenly commented:
> I would think that the tests should be able to steal from the way that the aliases for arguments are verified

I was thinking this, but I wasn’t sure quite how to implement it given that arg aliases are all strings, and I chose to implement this as a vector

> Let me be more explicit, by adding a way to do a similar thing but op returns a vector<string>, or by changing all of them to use vector<string> and just having the current ones wrap the single string in a vector.

I can fairly easily add a method to get all of the aliasFullNames as a vector, and create a method for `EnsureVectorStringsAreLowercaseAndNoCollisions` . Thank you for the advice!

denelon commented:
I've seen plenty of feedback in this area with the list of Issues. I think it will help in terms of muscle memory for users coming from other package managers.

Can we get a table of the original command and the aliases in the description? That would help with consistency when release notes come out.

How are aliases to be discovered?

Maybe: 
```text
winget search -?                                                                in pwsh at 16:56:34
Windows Package Manager v1.3.2091
Copyright (c) Microsoft Corporation. All rights reserved.

Searches for packages from configured sources.

usage: winget search [[-q] <query>] [<options>]

The following command aliases are available:
  find

The following arguments are available:
  -q,--query                  The query used to search for a package

The following options are available:
  --id                        Filter results by id
  --name                      Filter results by name
  --moniker                   Filter results by moniker
  --tag                       Filter results by tag
  --command                   Filter results by command
  -s,--source                 Find package using the specified source
  -n,--count                  Show no more than specified number of results (between 1 and 1000)
  -e,--exact                  Find package using exact match
  --header                    Optional Windows-Package-Manager REST source HTTP header
  --accept-source-agreements  Accept all source agreements during source operations

More help can be found at: https://aka.ms/winget-command-search
```

I expect these would only impact the CLI and not the PowerShell cmdlets as those have a discrete set of approved nouns and verbs.

JohnMcPMS commented:
Agreed, no change to PowerShell.

Trenly commented:
Agreed with all points

Trenly commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -183,6 +200,37 @@ namespace AppInstaller::CLI
+            for (const auto& commandAlias : commandAliases)
+            {
+                infoOut << Execution::HelpCommandEmphasis << commandAlias;
+                /*
+                *  Print up to 7 command aliases per line, comma separated for readability
+                *  This is intended to be future-looking so that in case common typos are
+                *  added to command aliases as scoop/chocolatey do, the list does not become
+                *  extremely long to the point where it makes the help page less usable.
+                */
```
For your reference @denelon -
![image](https://user-images.githubusercontent.com/12611259/181798253-68576bf5-d1b0-45a9-83e2-454d38dfcfb4.png)

Trenly commented:
> There has been a lack of rigor (mostly from me) about completion, with many new features not being added. I'm not going to hold you to it, but now that I've started thinking about it I see it more. - @JohnMcPMS 

I added command aliases to completion

> How are aliases to be discovered? - @denelon 

Added to the command help per the proposal

> Can we get a table of the original command and the aliases in the description? That would help with consistency when release notes come out. - @denelon 

| Original Command | New Aliases |
| ------------- | ------------- |
| search  | find  |
| install  | add |
| upgrade | update |
| uninstall | remove, rm |
| source update | source refresh |
| source remove | source rm |

lechacon commented:
Pie in the sky idea: What if instead of adding the aliases, we add a feature where users can create their own aliases, like git does?

denelon commented:
I could see something like that as a part of the settings.json file. It's an interesting idea. I'd expect we would need to be fairly strict when defining those aliases, but I like it.

Trenly commented:
> Pie in the sky idea: What if instead of adding the aliases, we add a feature where users can create their own aliases, like git does?

I think that having that option as a user setting would be good; I think that a combination here would be best, where default aliases are specified as they are here, but if a user specified one that conflicted with a pre-defined alias, the user-defined alias would take precedence. But, I think overall it's an interesting idea and could be a good feature in addition to the standard aliases

vedantmgoyal2009 commented:
|Command|Alias|
|-|-|
|install|i|
|list|ls|
|show|view|
|settings|config|

lechacon commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -291,7 +339,10 @@ namespace AppInstaller::CLI
         {
-            if (Utility::CaseInsensitiveEquals(*itr, command->Name()))
+            if (
+                Utility::CaseInsensitiveEquals(*itr, command->Name()) ||
+                Utility::CaseInsensitiveContains(command->Aliases(), *itr)
```
I think at some point we send telemetry about what command was executed with its full name. I think it could be helpful to report the specific alias. If most people use the alias than the original command, maybe we could rename the command in a v2 to what people prefer.

Trenly replied:
Assuming you're referring to this - https://github.com/microsoft/winget-cli/blob/a56372ba05d8f4423cb7cf5be0b5c0bacb445ca3/src/AppInstallerCLICore/Command.cpp#L888

Just clarifying - All that would need to be done to enable telemetry around specific aliases would be to pass the aliasFullName to the telemetry logger instead of the command full name?

lechacon replied:
I think it would be more complicated. We'd probably need a new event, like `Logging::Telemetry().LogCommandAlias(...)` to avoid tripping any data processing we currently have that looks at the full name. Also, we'd need to have `CaseInsensitiveContains` tell us which alias matched. Now that I'm saying this, it seems too complicated and I'm not sure if it is needed..

lechacon commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -43,6 +44,21 @@ namespace AppInstaller::CLI
+        {
+            for (auto& alias : aliases)
+            {
+                std::string _aliasFullName;
```
Nit: I think the leading underscore is not consistent with the style in the rest of the project

lechacon commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -43,6 +44,21 @@ namespace AppInstaller::CLI
+                std::string _aliasFullName;
+                if (!parent.empty())
+                {
+                    _aliasFullName = parent;
```
What if the parent is also aliased?

Trenly replied:
Are you suggesting that the entire tree needs to be expanded to find all potential alias full names?

lechacon replied:
I'm suggesting that we need to give some thought to that case. Depending on how we use the full name, we could need to generate all possible full names, or just one, or maybe we don't need the alias full names at all.

I *think* we only use the full name for logging? If that's the case we probably don't need to compute the alias full names, but we'd need to verify.

Trenly replied:
I did have to use the full name to ensure that there weren't collisions between aliases and commands, given that aliases can be the same but not collide if they have different parents. Since the collision check is recursive, I don't think the full expansion is needed. By checking that `root/search` and `root/find` don't collide with any other commands, then any sub-commands underneath them will also not collide with commands in other branches. So, as long as there are no aliases that collide at the same level, checking only one branch of the sub-trees suffices as a check for all other branches within that path.

lechacon commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -43,6 +44,21 @@ namespace AppInstaller::CLI
+                    _aliasFullName += ParentSplitChar;
+                }
+                _aliasFullName += alias;
+                m_aliasFullNames.emplace_back(_aliasFullName);
```
`std::move` here

lechacon commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -26,11 +26,12 @@ namespace AppInstaller::CLI
 
     Command::Command(
         std::string_view name,
+        std::vector<std::string_view> aliases,
```
Maybe this should be `const &`?
Edit: Or `&&`

Trenly replied:
I'm not sure I quite follow what you mean

lechacon replied:
I think the type should be `const std::vector<std::string_view>&` or `std::vector<std::string_view>&&` to avoid creating a copy of the vector when the constructor is called.

lechacon commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -183,6 +200,37 @@ namespace AppInstaller::CLI
+                *  added to command aliases as scoop/chocolatey do, the list does not become
+                *  extremely long to the point where it makes the help page less usable.
+                */
+                if (i%7 != 0 && i!= commandAliases.size()) {
```
Nit: Braces in new line for consistency

lechacon commented on src/AppInstallerCLICore/Command.h:
```diff
@@ -78,7 +80,9 @@ namespace AppInstaller::CLI
         constexpr static char ParentSplitChar = ':';
 
         std::string_view Name() const { return m_name; }
+        std::vector<std::string_view> Aliases() const { return m_aliases; }
```
Return a `const &` to avoid creating a copy.

lechacon commented on src/AppInstallerCLICore/Commands/SearchCommand.h:
```diff
@@ -5,9 +5,10 @@
     struct SearchCommand final : public Command
     {
-        SearchCommand(std::string_view parent) : Command("search", parent) {}
+        SearchCommand(std::string_view parent) : Command("search", {"find"}, parent) {}
```
Nit: Add spaces around `"find"` for consistency

lechacon commented on src/AppInstallerCLICore/Command.h:
```diff
@@ -56,16 +56,18 @@ namespace AppInstaller::CLI
-            Command(name, parent, Command::Visibility::Show, Settings::ExperimentalFeature::Feature::None, groupPolicy) {}
-        Command(std::string_view name, std::string_view parent, Command::Visibility visibility, Settings::ExperimentalFeature::Feature feature) :
-            Command(name, parent, visibility, feature, Settings::TogglePolicy::Policy::None) {}
-        Command(std::string_view name, std::string_view parent, Command::Visibility visibility, Settings::ExperimentalFeature::Feature feature, Settings::TogglePolicy::Policy groupPolicy);
+            Command(name, {}, parent) {}
+        Command(std::string_view name,std::vector<std::string_view> aliases, std::string_view parent) :
+            Command(name, aliases, parent, Settings::ExperimentalFeature::Feature::None) {}
+        Command(std::string_view name,std::vector<std::string_view> aliases, std::string_view parent, Command::Visibility visibility) :
+            Command(name, aliases, parent, visibility, Settings::ExperimentalFeature::Feature::None) {}
+        Command(std::string_view name,std::vector<std::string_view> aliases, std::string_view parent, Settings::ExperimentalFeature::Feature feature) :
+            Command(name, aliases, parent, Command::Visibility::Show, feature) {}
+        Command(std::string_view name,std::vector<std::string_view> aliases, std::string_view parent, Settings::TogglePolicy::Policy groupPolicy) :
+            Command(name, aliases, parent, Command::Visibility::Show, Settings::ExperimentalFeature::Feature::None, groupPolicy) {}
+        Command(std::string_view name,std::vector<std::string_view> aliases, std::string_view parent, Command::Visibility visibility, Settings::ExperimentalFeature::Feature feature) :
+            Command(name, aliases, parent, visibility, feature, Settings::TogglePolicy::Policy::None) {}
+        Command(std::string_view name,std::vector<std::string_view> aliases, std::string_view parent, Command::Visibility visibility, Settings::ExperimentalFeature::Feature feature, Settings::TogglePolicy::Policy groupPolicy);
```
Why not keep the existing constructors so that we don't need to pass the empty list for aliases when there aren't any?

Trenly replied:
I chose not to keep the existing constructors to force consideration of aliases and to reduce the amount of overloads. If the old constructors are still desired, I can add them back, but I thought it made more sense to just add aliases, or the empty vector, to everything.

lechacon commented on src/AppInstallerCLITests/Command.cpp:
```diff
@@ -47,6 +57,11 @@ std::string GetArgumentAlias(const Argument& arg)
 
+bool StringIsLowercase(std::string& s)
+{
+    return { Utility::ToLower(s) == s };
```
Nit: Are the braces needed?

lechacon commented on src/AppInstallerCLITests/Command.cpp:
```diff
@@ -47,6 +57,11 @@ std::string GetArgumentAlias(const Argument& arg)
     }
 }
 
+bool StringIsLowercase(std::string& s)
```
`const std::string&`?

lechacon commented on src/AppInstallerCommonCore/AppInstallerStrings.cpp:
```diff
@@ -105,6 +105,18 @@ namespace AppInstaller::Utility
 
+    bool CaseInsensitiveContains(std::vector<std::string_view> a, std::string_view b)
+    {
+        for (std::string_view c : a)
+        {
+            if (CaseInsensitiveEquals(c, b))
+            {
+                return true;
+            }
+        }
```
Not saying that it would be better, but FYI you can do this with `std::any_of`. Something like
```c++
return std::any_of(a.begin(), a.end(), [=](const std::string_view s) { return CaseInsensitiveEquals(s, b); });
```

wmstack commented:
I suggest that https://github.com/microsoft/winget-cli/issues/2302 is considered as well. I feel like user scope and machine scope could also be aliases when installing a package

wmstack commented:
If you want to find aliases, I think NPM does a great job with their aliases. They make it so that if you make a typo like `instal` or `istnall` it would still work. These could be *hidden* aliases:
```
npm install [<package-spec> ...]

aliases: add, i, in, ins, inst, insta, instal, isnt, isnta, isntal, isntall
```
> https://docs.npmjs.com/cli/v8/commands/npm-install


That might be considered butchering the aliases, lol. Otherwise we could have this mechanism that says "did you mean ...", however, I think most of the time the author really means it.

Another type of alias that I want to mention is adding `s` to signify listing:
```
# equivalent to `winget list`
winget packages
```

Trenly commented:
> I suggest that #2302 is considered as well. I feel like user scope and machine scope could also be aliases when installing a package (with `-g` like NPM and `-u` to force a user scope package).

Those are flags, and not commands. Therefore, they fall outside the scope of a command alias and cannot be considered as part of this PR

JohnMcPMS commented on src/AppInstallerCLICore/Command.cpp:
```diff
@@ -26,11 +26,12 @@ namespace AppInstaller::CLI
         Settings::ExperimentalFeature::Feature feature,
         Settings::TogglePolicy::Policy groupPolicy) :
-        m_name(name), m_visibility(visibility), m_feature(feature), m_groupPolicy(groupPolicy)
+        m_name(name), m_aliases(aliases), m_visibility(visibility), m_feature(feature), m_groupPolicy(groupPolicy)
```
```suggestion
        m_name(name), m_aliases(std::move(aliases)), m_visibility(visibility), m_feature(feature), m_groupPolicy(groupPolicy)
```
This solves @lechacon 's comment above.

JohnMcPMS commented on src/AppInstallerCommonCore/AppInstallerStrings.cpp:
```diff
@@ -105,6 +105,11 @@ namespace AppInstaller::Utility
 
+    bool CaseInsensitiveContains(std::vector<std::string_view> a, std::string_view b)
+    {
+        return std::any_of(a.begin(), a.end(), [=](const std::string_view s) { return CaseInsensitiveEquals(s, b); });
```
```suggestion
        auto B = ToLower(b);
        return std::any_of(a.begin(), a.end(), [&](const std::string_view& s) { return ToLower(s) == B; });
```

JohnMcPMS commented on src/AppInstallerCommonCore/AppInstallerStrings.cpp:
```diff
@@ -105,6 +105,11 @@ namespace AppInstaller::Utility
         return ToLower(a) == ToLower(b);
     }
 
+    bool CaseInsensitiveContains(std::vector<std::string_view> a, std::string_view b)
```
```suggestion
    bool CaseInsensitiveContains(const std::vector<std::string_view>& a, std::string_view b)
```

Trenly commented on src/AppInstallerCLITests/Command.cpp:
```diff
@@ -80,9 +94,41 @@ void EnsureStringsAreLowercaseAndNoCollisions(const std::string& info, const Enu
 {
-    EnsureStringsAreLowercaseAndNoCollisions(command.FullName() + " commands", command.GetCommands(), GetCommandName);
+    // Command names and aliases exist in the same space, so both need to be checked as a set
+    // However, collisions do not occur between levels, so the full name must be used to check for collision
```
Ahh, I see. I was misunderstanding

JohnMcPMS commented:
/azp run

Trenly commented:
I think I missed a type conversion in the test; I don't think std::string_view can be implicitly cast to std::string; If the build fails, I'll push a fix

JohnMcPMS commented:
/azp run
